Accept params as dictionary (default) or list (optional)

Stefan-Heimersheim opened the pull request:
This is useful for MCMC samplers and similar tools that pass
`params` as list instead of dictionary.
Note that `log_unnormalized_likelihood` is the only function that allows
for list of parameters. The list gets converted there and calls the
other functions. This avoids needing to convert this everywhere.

Resolves issue #7

Stefan-Heimersheim commented:
Are these tests supposed to pass or not implemented yet? [@steven-murray?] I.e. should we merge anyway at the moment?

steven-murray commented:
@Stefan-Heimersheim yes the tests should pass... you should do `pip install pre-commit` then `pre-commit install`. That will ensure that linting passes on your own system before letting you actually commit. I'll make a commit to try and fix the other tests.

steven-murray commented:
OK I think I've fixed it -- but the function looks quite incomplete. We should probably make the function complete before merging. Also, it may make sense to have what you've put into the function so far (converting list to dict) as its own hidden method, which would eventually be called by `log_unnormalized_likelihood`. 

FWIW, I prefer a shorter method name, like `log_lk` or `logl` or `log_likelihood` :-)

Stefan-Heimersheim commented:
Thanks! I was a bit confused because I didn't change the parts of the code the tests were complaining about.

> the function looks quite incomplete

Yes, it does. But that's the function from the master branch, I just added those lines ;)

> (converting list to dict) as its own hidden method

Sounds like a good idea, I'd say we can export this into it's own method when we are making the function more complete later.

steven-murray commented:
Yup, the deploy to PyPI won't work. In fact, maybe we should remove it. I put it in for completeness. 

I suggest moving the code you wrote in to its own complete method, and sticking a "NotImplementedError" into the likelihood function for now. Then it's a complete unit of work to merge.

steven-murray commented on src/pspec_likelihood/likelihood.py:
```diff
@@ -250,22 +249,21 @@ def params_to_dict(self, params):
-                type(params) is list
-            ), ("params is not a list, but params_list was given. "
-                "When params is a dictionary, leave params_list set to None.")
+            assert type(params) is list, (
```
I'm actually not sure we need to do this. If the user passes a dict for params, we can just use it without error.

Stefan-Heimersheim commented:
> I think you should also add the `params_list` to the class instance via the init. Although, I think it would be better called `param_names`.

My idea was to use `params_list` both as indicator if the user wants to pass a list, and as variable for the list. (I think when someone passes the parameter list it's obvious that they want to use the list-version.)
`params_list` is set in the ini (as kwarg).

    assert type(params) is list, ( 

This is in case the user gives us a list of parameters to pass *and* a dictionary. I thought something is clearly going wrong and we should raise an error, but other options would be a) Just assume the dict is correct and ignore the input. Or b) Check if dict.keys() == list (modulo order) and then ignore the input.

steven-murray commented:
Perhaps I'm misunderstanding, but the way I'm thinking of it working is like this:

```python
class Likelihood:
    def __init__(self, param_names=None):
        assert isinstance(param_names, [None, list, tuple])
        self.param_names = param_names

    def params_to_dict(self, params):
        if isinstance(params, dict):
            if self.param_names is not None:
                assert set(self.param_names) == set(params.keys()), "input parameters don't match parameters of the likelihood"
            return params
        else:
            if self.param_names is None:
                raise ValueError("To pass params as a sequence, rather than dict, likelihood must be created with param_names")    
        return {k:v for k,v in zip(self.param_names, params)}
```

In this case, if `param_names` is not given when initing, one can't use a list/tuple for `params`. If it _is_ given, one can use _either_ a dict or list for `params`. If a dict, it has to have the same keys as the definition `param_names`. But it's still able to be used. What might be even _better_ than this is to have a set of _fiducial_ parameters defined at instantiation, matching the `param_names`. If the user passes a dict as `params` then, it need only have a _subset_ of the keys, with other keys getting their fiducial value.

Stefan-Heimersheim commented:
I see what you meant, I changed the implementation of `params_to_dict` to reflect this (and also the name). Also your `{k:v for k,v in zip(self.param_names, params)}` version seems more elegant than my for loop -- does it work if params is either a list or tuple I assume? I added a check for that.

steven-murray commented:
@Stefan-Heimersheim yes the `zip` will work for any sequence (list/tuple/generator). It is odd that some of the checks are no longer showing :-/ Sometimes github does break... they'll probably come back soon :-)

Stefan-Heimersheim commented:
Thanks Steven for doing the merge already!

Not sure why the remaining "Tests/Testing"-tests are failing (different error messages, always "Interrupted")
```
 =========================== short test summary info ============================
ERROR tests/test_basic.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.15s ===============================
Error: Process completed with exit code 2.
```

Otherwise I think we can merge this into master then, can we?
